storage/innobase/buf/buf0buf.cc

```diff
@@ -2235,29 +2240,33 @@ buf_page_t *buf_page_get_zip(const page_id_t page_id) noexcept
     if (!bpage)
     {
       hash_lock.unlock_shared();
-      switch (dberr_t err= buf_read_page(page_id, chain, false)) {
-      case DB_SUCCESS:
-      case DB_SUCCESS_LOCKED_REC:
-        mariadb_increment_pages_read(stats);
-        continue;
-      case DB_TABLESPACE_DELETED:
-        return nullptr;
-      default:
-        sql_print_error("InnoDB: Reading compressed page "
-                        "[page id: space=" UINT32PF ", page number=" UINT32PF
-                        "] failed with error: %s",
-                        page_id.space(), page_id.page_no(), ut_strerr(err));
+      dberr_t err;
+      bpage= &buf_read_page(page_id, &err, chain, false)->page;
+      if (!bpage)
+      {
+        if (err != DB_TABLESPACE_DELETED)
+          sql_print_error("InnoDB: Reading compressed page "
+                          "[page id: space=" UINT32PF ", page number=" UINT32PF
+                          "] failed with error: %s",
+                          page_id.space(), page_id.page_no(), ut_strerr(err));
         return nullptr;
       }
+      hash_lock.lock_shared();
+      bpage->unfix();
     }
 
-    ut_ad(bpage->in_file());
+    ut_d(uint32_t state= bpage->state());
+    ut_ad(state >= buf_page_t::UNFIXED);
```

Comment:
what if freed state page exist in buffer pool?  This can be easily accessed during mvcc lookup. I think this assert is wrong.

Reply:
6e821dd0c96970727e1dec1b71800f4e52f33baf reverts this change.

In [MDEV-37753](https://jira.mariadb.org/browse/MDEV-37753) I just wrote that an assertion needs to be added to the end of the function, to cover also the case where the page will have to be read into the buffer pool. This will be done as part of implementing that fix.